runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.nonrepudiation.api.v1
+
+import java.io.ByteArrayInputStream
+import java.security.cert.{CertificateFactory, X509Certificate}
+
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.{Route, StandardRoute, ValidationRejection}
+import com.daml.nonrepudiation.{CertificateRepository, FingerprintBytes}
+import com.google.common.io.BaseEncoding
+import org.slf4j.{Logger, LoggerFactory}
+import spray.json.{DefaultJsonProtocol, RootJsonFormat}
+
+import scala.util.Try
+
+private[api] final class CertificatesEndpoint private (certificates: CertificateRepository)
+    extends CertificatesEndpoint.JsonProtocol {
+
+  import CertificatesEndpoint._
+
+  private def putCertificate(certificate: X509Certificate): Route =
```

Comment:
what happens if I add the same certificate twice?

Reply:
Good question. My intention was to be idempotent but right know it's likely to fail on the backend, I'll add a test for this and fix it.

Reply:
This requires a few changes to the PostgreSQL part, I'll merge this as is and open a separate PR to guarantee that the back-end guarantees idempotency.

Reply:
Tracked in https://github.com/digital-asset/daml/issues/9021